Variable Relocation for Various Models using Radiant Style Updating

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #9926 
 - Numerous models that have an impact simultaneously in the heat balances of the zone and the HVAC simulation use a strategy that is similar to what was developed for the low temperature radiant systems.  This allows the entity to simulate at the HVAC time step to look at how the system responds on that time step while still allowing it to keep in synch with the zone heat balances which could be running at a different time step.  These models all followed a strategy which deposited information in a higher level data structure based on how things were when F90 was used in the code.  As we look to improve the code and follow more contained models, there was a desire to move all appropriate variables to local data structures.  The main variable that was tagged for improvement went by the name of ZeroSourceSumHAT which calculated the zone conditions when a particular system or model was off and then compared it to an updated calculation for SumHAT to determine the potential impact of the system on the zone conditions beyond just what was added to the zone air convectively directly from the system.  While not perfect, it allows all aspects of these model to impact what is happening and what might be needed from other HVAC systems to meet the proper conditions.  This includes radiant convective baseboards models, low temperature radiant systems, high temperature radiant systems, swimming pools, simple chilled ceiling panels, and ventilated slabs.  In the case of the low temperature radiant systems and the ventilated slabs, a slight variation on the new strategy was needed because these two systems have terms in the surface heat balances due to the fact that these heat sources/sinks are embedded in a particular surface.  This work was complete and also all data is now local following work that was done previously on a limited basis for one of these systems.  This work includes a unit test which specifically looks at one of the special cases (low temperature radiant systems).  The test suite results only show small changes in one model--the simple chilled ceiling panel model.  The reason for the minor changes in that model is because during this work a minor bug was uncovered and fixed.  All other files show no differences in the output as expected.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Correction of Issues with ZeroSource variables

Changes made to ZeroSource variables in various models that act like radiant systems.  ZeroSource (and other variables) converted from arrays over zones or surfaces to being attached to local objects.

RKStrand committed:
Corrections to Unit Tests

Corrections made to HW Baseboard Radiator unit tests which used one of the variables that was moved to become a local variable.

RKStrand committed:
Merge branch 'develop' into 9926ZeroSourceTermCorrections

RKStrand committed:
More corrections

Additional changes needed after attempting to merge in some diffs with develop.

RKStrand committed:
Changes to Address Diffs and Fails

Fixes a bug introduced in some of the changes that were made and tries to address one of the models that had significant differences in the output.

RKStrand committed:
Another Round of Corrections

Corrections made to address large differences in the results for several idfs and also to take care of some warning messages from the compiler.

RKStrand committed:
Correction of Source to Avg Assignment

Rewrite potentially lost some data transfer when one high temperature radiant heater was on--it broke out of the for loop and missed the remaining assignments.  This is corrected here.

RKStrand committed:
Clean up swimming pool

Cudos to @rraustad for catching this before it became a new bug.

RKStrand committed:
Correction to HTR

Got overly aggressive with a change, this impacted some initializations in the high temperature radiant heater (HTR).

RKStrand committed:
Merge branch 'develop' into 9926ZeroSourceTermCorrections

RKStrand committed:
Swimming Pool Unit Test and Steam Baseboard Fix

Fixed some issues with the swimming pool and the steam baseboard that caused problems in the runs.

RKStrand committed:
Additional changes to certain models

Additional changes specific to models that impact the surface heat balances, requiring a slightly different approach.

RKStrand committed:
Merge branch 'develop' into 9926ZeroSourceTermCorrections

RKStrand committed:
Minor updates

Got rid of unused variables causing warnings and made another initialization to make sure just is not being left around in an array.

RKStrand committed:
Minor typo fix

No comment.

RKStrand committed:
Unit Test for ZeroSource Fix

Addition of a unit test for the fixes for various ZeroSource variables throughout the code.

RKStrand committed:
Merge branch 'develop' into 9926ZeroSourceTermCorrections

RKStrand committed:
Clang formating issue

Fixed something that clang didn't like.

rraustad commented on src/EnergyPlus/HWBaseboardRadiator.hh:
```diff
@@ -113,22 +113,42 @@ namespace HWBaseboardRadiator {
+        Real64 QBBRadSrcAvg;       // Need to keep the last value in case we are still iterating
+        Real64 LastSysTimeElapsed; // Need to keep the last value in case we are still iterating
+        Real64 LastTimeStepSys;    // Need to keep the last value in case we are still iterating
+        Real64 LastQBBRadSrc;      // Need to keep the last value in case we are still iterating
```
There are several things here. 1 is that ZeroBBSourceSumHATsurf, QBBRad* variables were moved to the HWBB struct. 2 the Last* variables were also moved to the same struct without deleting the AccountingData struct. 3 this code used to initialize inline and here a constructor was added back in. We are moving away from constructors and initializing in line where the variable is decalred.

RKStrand replied:
I'll look at this again, hopefully later this afternoon.  If I get rid of the stuff in the Accounting data structure and go back to inline initializations, would that address your concerns here?

rraustad replied:
Yes, I don't have a hard opinion on which method is better. I do like the grouping approach for like variables but this can be done either way.

RKStrand replied:
I went with the approach that was more like all the other models that do this--without the AccountingData struct.  So, I got rid of the redundant parameters and that struct and move all of the initializations to in line, eliminating the constructor completely.  Committing this momentarily...

rraustad commented on src/EnergyPlus/HighTempRadiantSystem.hh:
```diff
@@ -198,14 +206,7 @@ struct HighTempRadiantSystemData : BaseGlobalStruct
-    Array1D<Real64> LastTimeStepSys;    // Need to keep the last value in case we are still iterating
+    int NumOfHighTempRadSys = 0; // Number of hydronic low tempererature radiant systems
     Array1D_bool MySizeFlag;
     Array1D_bool CheckEquipName;
```
Moving these variables cleaned up a lot of local arrays.

rraustad commented on src/EnergyPlus/HighTempRadiantSystem.cc:
```diff
@@ -1422,22 +1416,22 @@ namespace HighTempRadiantSystem {
         Real64 TimeStepSysSec = state.dataHVACGlobal->TimeStepSysSec;
```
This would help here:  `auto &thisHTR = state.dataHighTempRadSys->HighTempRadSys(RadSysNum);`

RKStrand replied:
I'll definitely go back and add that!

RKStrand replied:
Implemented this in a variety of subroutines in this module that hadn't used it before.  Much cleaner now.  Will show up in the next commit.

RKStrand replied:
Did an intermediate commit specifically for this change so as to keep change easy to separate if something goes awry.

rraustad commented on src/EnergyPlus/LowTempRadiantSystem.cc:
```diff
@@ -1972,34 +1971,53 @@ namespace LowTempRadiantSystem {
+                thisLTR.LastQRadSysSrc.dimension(thisLTR.NumOfSurfaces, 0.0);
+                thisLTR.LastSysTimeElapsed.dimension(thisLTR.NumOfSurfaces, 0.0);
+                thisLTR.LastTimeStepSys.dimension(thisLTR.NumOfSurfaces, 0.0);
+            }
```
Why are these variables in the LowTempRadiantSystem allocated to the number of surfaces while the HIghTempRadiantSystem holds a scalar?

rraustad replied:
Note here that LastSysTimeElapsed and LastTimeStepSys are arrays which all elements would hold the same value. These could be scalars like all the others (better think through that logic before changing it)? I think I also saw this in VentilatedSlabs so check there as well.

RKStrand replied:
Hmmm...yeah, I'll have to think about that a little more.  I am pretty sure that the time quantities do not need to be arrays.  Those are going to be set at the system level, not at the surface level.  I'll go back and look at the LTR and Ventilated Slab and think that through some more and then make changes as necessary.  Hopefully I can look at this tomorrow some more after I finish up another defect for which I am in the midst of putting a unit test together.  Thanks again for your input on this!

RKStrand replied:
@rraustad Yup, you are absolutely correct here.  These did not need to be arrays.  I've fixed this for both the low temperature radiant system and the ventilated slab and made a commit just now.  There are still some other things I need to look at, but I wanted to make a commit specifically for this so that it was separate from any other changes I make that might not end up being as clean potentially.

rraustad replied:
3 object loops in a row.

rraustad replied:
-3

rraustad commented on src/EnergyPlus/SteamBaseboardRadiator.hh:
```diff
@@ -209,15 +220,6 @@ struct SteamBaseboardRadiatorData : BaseGlobalStruct
-    // Record keeping variables used to calculate QBBRadSrcAvg locally
-    Array1D<Real64> LastQBBSteamRadSrc; // Need to keep the last value in case we are still iterating
-    Array1D<Real64> LastSysTimeElapsed; // Need to keep the last value in case we are still iterating
-    Array1D<Real64> LastTimeStepSys;    // Need to keep the last value in case we are still iterating
```
And here these used to be arrays and now they are scalars?

RKStrand replied:
Yes, only a single Q term to deal with because there isn't a separate heat source that is unique for each surface like in low temperature radiant and ventilated slab.

rraustad reviewed (commented):
I am confused as to why some models use arrays for the variables in question and others use scalars.

RKStrand commented:
@rraustad I'll try to clarify why some of the models use scalar and some use arrays.  Two of the models (low temperature systems and ventilated slab) have the potential to have multiple surfaces associated with a single system.  As a result, those need arrays for anything related to a heat flux.  I think I only did those variables and not the time step variables and only for those two models.  Everything else only needed to capture a single value rather than one for multiple surfaces.  Also, in case you are wondering, one of the variables stayed at the higher level because it is used in the heat balance equations directly. I don't think there is a way around that unless we implement some potentially awkward call to the local model from the heat balance and search for the value.  I think that would be slow because you'd have to figure out which system that surface is in.  Does this help at all?  I think overall there is a lot less data and almost all of the data is now local.

RKStrand committed:
Merge branch 'develop' into 9926ZeroSourceTermCorrections

RKStrand committed:
Merge with Corrections

Merged in latest develop and cleaned up some other stuff that somehow made it through previously (???).

RKStrand committed:
clang format didn't like it

clang format detected a double blank line and said nope.  Fixed here.

RKStrand committed:
LTR and VentSlab time variables

During review, it was noted that the time variables for the LTR and VentSlab systems did not need to be arrays but rather a scalar for a specific system.  This commit fixes that issue.

RKStrand committed:
More intros of auto statements

Implementation of more "thisHTR" auto definitions to clean up the code in various subroutines throughout this module.

RKStrand committed:
AccountingData fix and random HTR fix

Corrected the data structure for the HWBB based on review comments.  Also uncovered a problem in HTR that somehow slipped by the compiler on a previous build (?).

RKStrand committed:
Merge branch 'develop' into 9926ZeroSourceTermCorrections

RKStrand committed:
Merge branch 'develop' into 9926ZeroSourceTermCorrections

nrel-bot-2b commented:
@RKStrand it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@RKStrand it has been 7 days since this pull request was last updated.

nrel-bot-2b commented:
@RKStrand it has been 12 days since this pull request was last updated.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 9926ZeroSourceTermCorrections

mjwitte commented on src/EnergyPlus/ChilledCeilingPanelSimple.cc:
```diff
@@ -889,7 +889,7 @@ void InitCoolingPanel(EnergyPlusData &state, int const CoolingPanelNum, int cons
     if (state.dataGlobal->BeginTimeStepFlag && FirstHVACIteration) {
         int ZoneNum = thisCP.ZonePtr;
-        state.dataHeatBal->Zone(ZoneNum).ZeroSourceSumHATsurf = state.dataHeatBal->Zone(ZoneNum).sumHATsurf(state);
+        thisCP.ZeroCPSourceSumHATsurf = state.dataHeatBal->Zone(ZoneNum).sumHATsurf(state);
```
Array to scalar 👍

mjwitte commented on src/EnergyPlus/HWBaseboardRadiator.cc:
```diff
@@ -848,17 +847,21 @@ namespace HWBaseboardRadiator {
             state.dataHWBaseboardRad->MyOneTimeFlag = false;
 
             for (int Loop = 1; Loop <= NumHWBaseboards; ++Loop) {
+                auto &hWBB = state.dataHWBaseboardRad->HWBaseboard;
```
Lots of new shortcut references.

mjwitte commented on src/EnergyPlus/HWBaseboardRadiator.hh:
```diff
@@ -206,9 +204,6 @@ namespace HWBaseboardRadiator {
 struct HWBaseboardRadiatorData : BaseGlobalStruct
 {
 
-    Array1D<Real64> QBBRadSource;         // Need to keep the last value in case we are still iterating
-    Array1D<Real64> QBBRadSrcAvg;         // Need to keep the last value in case we are still iterating
-    Array1D<Real64> ZeroSourceSumHATsurf; // Equal to the SumHATsurf for all the walls in a zone with no source
```
Loose arrays moved into the main struct. 👍

mjwitte reviewed (approved):
@RKStrand Lots of great array reductions and other code cleanup here. I've merged in develop so CI can run another pass.
@rraustad It's been a while since you commented on this. Any further review you want to do?

rraustad commented on src/EnergyPlus/HWBaseboardRadiator.cc:
```diff
@@ -1501,35 +1499,31 @@ namespace HWBaseboardRadiator {
 
-        for (BaseboardNum = 1; BaseboardNum <= NumHWBaseboards; ++BaseboardNum) {
-            auto &HWBaseboard = state.dataHWBaseboardRad->HWBaseboard(BaseboardNum);
+        for (BaseboardNum = 1; BaseboardNum <= state.dataHWBaseboardRad->NumHWBaseboards; ++BaseboardNum) {
```
Learning opportunity. I searched/highlighted, on this page, for BaseboardNum. The only occurrence is at line 1508. So then this could be:

    for (auto &thisHWBB : state.dataHWBaseboardRad->HWBaseboard)

rraustad commented on src/EnergyPlus/HighTempRadiantSystem.cc:
```diff
@@ -634,28 +634,27 @@ namespace HighTempRadiantSystem {
                                        "It will not be simulated.",
-                                       state.dataHighTempRadSys->HighTempRadSys(Loop).Name));
+                                       state.dataHighTempRadSys->HighTempRadSys(HTRnum).Name));
             }
```
Looks like another opportunity for an object loop.

rraustad replied:
-1

rraustad commented on src/EnergyPlus/HighTempRadiantSystem.cc:
```diff
@@ -666,29 +665,31 @@ namespace HighTempRadiantSystem {
+                thisHTR.LastQHTRRadSrc = 0.0;
+                thisHTR.LastSysTimeElapsed = 0.0;
+                thisHTR.LastTimeStepSys = 0.0;
+            }
```
Object loop here also. They are sometimes hard to see, this one is easy.

rraustad replied:
+1

rraustad commented on src/EnergyPlus/HighTempRadiantSystem.cc:
```diff
@@ -1094,19 +1098,17 @@ namespace HighTempRadiantSystem {
+            auto &thisHTR = state.dataHighTempRadSys->HighTempRadSys(RadSysNum);
+            thisHTR.QHTRRadSource = thisHTR.QHTRRadSrcAvg;
+            if (thisHTR.QHTRRadSrcAvg != 0.0) HighTempRadSysOn = true;
         }
```
Another example, where do you see loop counter RadSysNum, only in the creation of the reference.

rraustad replied:
+1

rraustad commented on src/EnergyPlus/HighTempRadiantSystem.cc:
```diff
@@ -1136,43 +1138,37 @@ namespace HighTempRadiantSystem {
+        dataHBFS->SurfQHTRadSys = 0.0;
+        dataHBFS->ZoneQHTRadSysToPerson = 0.0;
 
         for (int RadSysNum = 1; RadSysNum <= state.dataHighTempRadSys->NumOfHighTempRadSys; ++RadSysNum) {
```
Another opportunity for object loop. You cleaned this up so much it stands out now. Now you get the point of what to look for. I am not by any means the expert but this is what I look for during code review.

rraustad commented on tst/EnergyPlus/unit/LowTempRadiantSystem.unit.cc:
```diff
@@ -4663,3 +4663,135 @@ TEST_F(LowTempRadiantSystemTest, VariableFlowCoolingOnlyInputTest)
+    EXPECT_NEAR(qRadSysSource(4), 300.0, errTol);
+    EXPECT_NEAR(qRadSysSource(5), 400.0, errTol);
+    EXPECT_NEAR(qRadSysSource(6), 500.0, errTol);
+}
```
Seems well tested. Is a little hard on the eyes to follow, but I do like the unique numerics to track what's happening.

rraustad reviewed (approved):
Nice job @RKStrand.

RKStrand commented:
@mjwitte @rraustad Thanks for your comments and suggestions!  I have update the object loops that were noted above (see next commit once I have done a little testing locally).

rraustad commented:
The diffs I looked at don't seem to be an issue.

rraustad commented:
Here's the speed test before your next commit.

![image](https://github.com/NREL/EnergyPlus/assets/5778156/9d3d0089-0ff5-4490-9ca8-617d0947f068)

RKStrand committed:
I object (loop)!

A few more changes to address comments.

rraustad commented on src/EnergyPlus/LowTempRadiantSystem.cc:
```diff
@@ -2273,11 +2291,28 @@ namespace LowTempRadiantSystem {
+                thisLTR.LastQRadSysSrc = 0.0;
+                thisLTR.LastSysTimeElapsed = 0.0;
+                thisLTR.LastTimeStepSys = 0.0;
+            }
```
👀

rraustad commented on src/EnergyPlus/VentilatedSlab.cc:
```diff
@@ -1539,19 +1538,19 @@ namespace VentilatedSlab {
+                thisVentSlab.LastQRadSysSrc.dimension(numRadSurfs, 0.0);
+                thisVentSlab.LastSysTimeElapsed = 0.0;
+                thisVentSlab.LastTimeStepSys = 0.0;
             }
```
not sure

RKStrand replied:
@rraustad I was confused by the numbers...what were they supposed to indicate?  In any case, I went through the stuff that you fixed...all pretty solid stuff.  Shocking that one section of code just got axed and apparently was never needed?  Thanks for your thorough review and getting in the trenches on this one.

rraustad replied:
@RKStrand I ran local CppCheck on this code and it pointed me to variables that were initialized and never used. That's how that chunk of code got axed.

rraustad reviewed (commented):
OK

rraustad committed:
Object loop, unused vars, cleanup with CppCheck help

rraustad commented:
@Myoldmopar are the sprintf warnings on Mac a known issue?

Myoldmopar commented:
> @Myoldmopar are the sprintf warnings on Mac a known issue?

Yep, some have already been fixed, @nealkruis will be fixing the one in penumbra, and I'll make a PR to clean the others soon.

nrel-bot-2b commented:
<a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f39990ecad2b0fa95668e5580c6b948cf6de3400-x86_64-Linux-Ubuntu-22.04-gcc-11.4.html'>Unhandled Failure</a>

nrel-bot-2b commented:
9926ZeroSourceTermCorrections (rraustad) - x86_64-Linux-Ubuntu-22.04-gcc-11.4: OK (3520 of 3539 tests passed, 762 test warnings)

Messages:\n
 * 781 tests had: AUD diffs.
 * 430 tests had: EIO diffs.
 * 220 tests had: MDD diffs.
 * 243 tests had: MTD diffs.
 * 143 tests had: BND diffs.
 * 4 tests had: ESO small diffs.
 * 3 tests had: MTR small diffs.
 * 2 tests had: Table small diffs.
 * 156 tests had: RDD diffs.
 * 3 tests had: PERF_LOG diffs.
 * 3 tests had: ERR diffs.
 * 2 tests had: EDD diffs.
 * 19 tests had: Table big diffs.

Failures:\n

regression Test Summary
 * Passed: 784
 * Failed: 19


<a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f39990ecad2b0fa95668e5580c6b948cf6de3400-x86_64-Linux-Ubuntu-22.04-gcc-11.4.html'>![Build Badge](http://img.shields.io/badge/build%20status-passing-green.png)</a> <a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f39990ecad2b0fa95668e5580c6b948cf6de3400-x86_64-Linux-Ubuntu-22.04-gcc-11.4.html'>![Test Badge](http://img.shields.io/badge/tests%20passed-99.46%25-yellow.png)</a>

nrel-bot-2b commented:
9926ZeroSourceTermCorrections (rraustad) - x86_64-Linux-Ubuntu-22.04-gcc-11.4-UnitTestsCoverage-Debug: OK (1930 of 1930 tests passed, 0 test warnings)






<a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f39990ecad2b0fa95668e5580c6b948cf6de3400-x86_64-Linux-Ubuntu-22.04-gcc-11.4-UnitTestsCoverage-Debug.html'>![Build Badge](http://img.shields.io/badge/build%20status-passing-green.png)</a> <a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f39990ecad2b0fa95668e5580c6b948cf6de3400-x86_64-Linux-Ubuntu-22.04-gcc-11.4-UnitTestsCoverage-Debug.html'>![Test Badge](http://img.shields.io/badge/tests%20passed-100.0%25-green.png)</a> <a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f39990ecad2b0fa95668e5580c6b948cf6de3400-x86_64-Linux-Ubuntu-22.04-gcc-11.4-UnitTestsCoverage-Debug.html'>![Coverage Badge](http://img.shields.io/badge/coverage%20status-50.32%25-green.png)</a>

nrel-bot-2b commented:
9926ZeroSourceTermCorrections (rraustad) - x86_64-Linux-Ubuntu-22.04-gcc-11.4-IntegrationCoverage-Debug: OK (787 of 787 tests passed, 0 test warnings)






<a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f39990ecad2b0fa95668e5580c6b948cf6de3400-x86_64-Linux-Ubuntu-22.04-gcc-11.4-IntegrationCoverage-Debug.html'>![Build Badge](http://img.shields.io/badge/build%20status-passing-green.png)</a> <a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f39990ecad2b0fa95668e5580c6b948cf6de3400-x86_64-Linux-Ubuntu-22.04-gcc-11.4-IntegrationCoverage-Debug.html'>![Test Badge](http://img.shields.io/badge/tests%20passed-100.0%25-green.png)</a> <a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f39990ecad2b0fa95668e5580c6b948cf6de3400-x86_64-Linux-Ubuntu-22.04-gcc-11.4-IntegrationCoverage-Debug.html'>![Coverage Badge](http://img.shields.io/badge/coverage%20status-68.58%25-green.png)</a>

Myoldmopar commented:
The big regression diffs showing up on CI were related to some things I was trying over the weekend, and I have since reverted back.  I think it's worth a fresh set, so I'm pulling develop in here.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 9926ZeroSourceTermCorrections

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 9926ZeroSourceTermCorrections

mjwitte commented:
This has 2 approvals and just needs some fresh CI results. Assuming it's all green, this will merge today.

mjwitte commented:
CI is all clean. This introduces a few small diffs, so marking this ready to review, will merge some no-diffs PRs before this one.
